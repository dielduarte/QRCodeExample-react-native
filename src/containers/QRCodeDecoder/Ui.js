import React, { Fragment } from 'react';
import QRdecoder from 'react-native-qrimage-decoder';

import Button from '../../components/Button';

const Ui = ({ source, showImagePicker, onSuccess, onError }) => {
  return (
    <Fragment>
      <Button title="Scan from gallery" onPress={showImagePicker} />
      <QRdecoder src={source} onSuccess={onSuccess} onError={onError} />
    </Fragment>
  );
};

export default Ui;
