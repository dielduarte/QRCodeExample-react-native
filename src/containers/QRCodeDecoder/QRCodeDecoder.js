/**
 * Isolate the bussines logic and Ui
 * I chosed recompose todo that in a easy way
 * in a real scenario, we can do it for example to:
 * 1 - use the bussines logic of the component on the web/mobile apps
 * without need to rewrite it.
 * 2 - keep the code more clean and testable.
 */

import { withHandlers, withState, compose } from 'recompose';
import ImagePicker from 'react-native-image-picker';

import Ui from './Ui';

export const getMimeType = uri => {
  const ext = uri.substr(uri.lastIndexOf('.')).toLowerCase();

  const mimeTypeExts = {
    '.jpg': 'jpeg',
    '.jpeg': 'jpeg',
    '.jpe': 'jpeg',
    '.png': 'png',
    '.gif': 'gif'
  };

  const mimetype = mimeTypeExts[ext];

  return mimetype ? `image/${mimetype}` : false;
};

export const isUriValid = uri => uri.lastIndexOf('.') > 0;

export const imagePickerCallBack = (props, response) => {
  if (response.didCancel) return;

  if (response.error) {
    console.log('ImagePicker Error: ', response.error);
    return;
  }

  const { uri, data } = response;

  if (isUriValid(uri)) {
    const mimetype = getMimeType(uri);

    if (mimetype) {
      const source = `data:${mimetype};base64,${data}`;
      props.setSource(source);
    }
  }
};

const QRCodeDecoder = compose(
  withState('source', 'setSource', ''),
  withHandlers({
    showImagePicker: props => () => {
      ImagePicker.showImagePicker(response => imagePickerCallBack(props, response));
    },
    onSuccess: () => data => {
      alert(data);
    },
    onError: () => data => {
      alert(data);
    }
  })
);

export default QRCodeDecoder(Ui);
