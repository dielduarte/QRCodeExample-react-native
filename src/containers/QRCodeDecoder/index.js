import QRCodeDecoder from './QRCodeDecoder';
export default QRCodeDecoder;
