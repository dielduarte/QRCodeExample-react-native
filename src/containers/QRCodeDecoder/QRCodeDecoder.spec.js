import React from 'react';
import { shallow } from 'enzyme';
import ImagePicker from 'react-native-image-picker';

ImagePicker.showImagePicker = jest.fn();

import QRCodeDecoder, { getMimeType, isUriValid, imagePickerCallBack } from './QRCodeDecoder';

describe('QRCodeDecoder', () => {
  it('should render correctly', () => {
    const component = shallow(<QRCodeDecoder />).dive();
    expect(component).toMatchSnapshot();
  });

  describe('imagePickerCallBack', () => {
    it(`should return false if response.didCancel isn't null or undefined`, () => {
      const response = {
        didCancel: true
      };

      expect(imagePickerCallBack({}, response)).toBeFalsy();
    });

    it(`should return false if response.error isn't null or undefined`, () => {
      const response = {
        error: true
      };

      expect(imagePickerCallBack({}, response)).toBeFalsy();
    });

    it(`should call setSource prop passing the correct base64 source`, () => {
      const response = { uri: 'path-to-image.jpeg', data: 'falseDatabase64==' };
      const props = { setSource: jest.fn() };
      const expectedValue = `data:image/jpeg;base64,${response.data}`;
      imagePickerCallBack(props, response);

      expect(props.setSource).toHaveBeenCalledWith(expectedValue);
    });
  });

  describe('getMimeType', () => {
    it('should return the correct mime type when getMimeType is called', () => {
      expect(getMimeType('path-to-image.jpeg')).toEqual('image/jpeg');
    });

    it('should return false when getMimeType is called passing a non allowed format as param', () => {
      expect(getMimeType('path-to-image.docx')).toBeFalsy();
    });
  });

  describe('isUriValid', () => {
    it('should return true if uri passed is valid', () => {
      expect(isUriValid('path-to-image.jpeg')).toBeTruthy();
    });

    it(`should return false if uri passed haven't extension`, () => {
      expect(isUriValid('path-to-image')).toBeFalsy();
    });
  });

  it('should call ImagePicker.showImagePicker when showImagePicker prop is called', () => {
    const component = shallow(<QRCodeDecoder />).dive();
    component.props().showImagePicker();
    expect(ImagePicker.showImagePicker).toHaveBeenCalled();
  });
});
