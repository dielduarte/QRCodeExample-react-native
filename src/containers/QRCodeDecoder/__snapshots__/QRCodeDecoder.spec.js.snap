// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QRCodeDecoder should render correctly 1`] = `
<Ui
  onError={[Function]}
  onSuccess={[Function]}
  setSource={[Function]}
  showImagePicker={[Function]}
  source=""
/>
`;
